bionic/decorators.py

```diff
@@ -387,6 +390,54 @@ def pyplot(name=None, savefig_kwargs=None):
     )
 
 
+def accepts(**descriptors_by_arg_name):
+    """
+    Indicates that some of the decorated function's arguments should be supplied
+    according to certain descriptors. Each keyword argument to this decorator
+    corresponds to an argument of the decorated function -- that argument's value
+    will correspond to the descriptor passed to the decorator.
+
+    For example:
+
+        @builder
+        @accepts(pair="x, y")
+        def f(pair, z):
+
+    `f` will be called with two arguments: the first (`pair`) will be a tuple
+    containing the values of entities `x` and `y`, and the second (`z`) will just be
+    the value of entity `z`.
+
+    This decorator is currently experimental and does not have any additional
+    user-facing documentation. It may change in non-backwards-compatible ways.
+    """
+
+    outer_dnodes_by_inner = {
```

Comment:
Yes, `inner` and `outer` is a convention used in several providers. I think `base` is a little misleading because the inner provider is not a base class of the outer one, and `dep` is misleading because the dependency relationship is orthogonal to the inner/outer one. That said, I'm open to revisiting this naming, but probably we should do it in a different PR?

Reply:
Makes sense to revisit this later.